onnxruntime/core/providers/nnapi/nnapi_builtin/builders/qdq_support_helper.h

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <string>
+#include "core/graph/basic_types.h"
+#include "core/providers/nnapi/nnapi_builtin/nnapi_lib/NeuralNetworksTypes.h"
+#include "core/optimizer/qdq_transformer/selectors_actions/qdq_selectors.h"
+#include "core/optimizer/selectors_actions/helpers.h"
+
+namespace onnxruntime {
+
+class GraphViewer;
+class Node;
+
+namespace nnapi {
+
+struct Selector {
+  using OpVersionsMap = std::unordered_map<std::string, std::vector<ONNX_NAMESPACE::OperatorSetVersion>>;
+
+  Selector(const OpVersionsMap& ops_and_versions_in,
+           std::unique_ptr<QDQ::BaseSelector> selector_in)
+      : op_versions_map{ops_and_versions_in},
+        selector{std::move(selector_in)} {}
+
+  OpVersionsMap op_versions_map;
+  std::unique_ptr<QDQ::BaseSelector> selector;
+
+  ORT_DISALLOW_COPY_AND_ASSIGNMENT(Selector);
+};
+
+class Selectors {
+ public:
+  Selectors() = default;
+
+  Selectors(Selectors&& rhs) noexcept
+      : selectors_set_{std::move(rhs.selectors_set_)} {}
+
+  void RegisterSelector(const Selector::OpVersionsMap& ops_and_versions_in,
+                        std::unique_ptr<QDQ::BaseSelector> selector_in);
+
+  const std::unordered_set<std::unique_ptr<Selector>>& SelectorsSet() const {
+    return selectors_set_;
+  }
+
+  ORT_DISALLOW_COPY_AND_ASSIGNMENT(Selectors);
+
+ private:
+  std::unordered_set<std::unique_ptr<Selector>> selectors_set_;
+};
+
+class QDQSupportHelper {
+ public:
+  QDQSupportHelper(Selectors&& selectors);
+
+  std::optional<QDQ::NodeGroup> Match(const GraphViewer& graph_viewer, const Node& node) const;
+
+  bool IsNodeInQDQGroup(const Node& node);
+
+  std::optional<QDQ::NodeGroup> GetQDQNodeGroup(const onnxruntime::GraphViewer& graph_viewer, const Node& node);
+
+  void GetQDQNodeGroups(const onnxruntime::GraphViewer& graph_viewer);
+
+  Selectors selectors_;
+
+  std::vector<std::optional<QDQ::NodeGroup>> qdq_node_groups_;
+
+  std::unordered_map<const Node*, std::optional<QDQ::NodeGroup>> target_node_to_qdq_group_;
+
+  std::unordered_map<std::string, const Selector*> op_type_to_selectors_map_;
+
+  std::vector<const Node*> dq_nodes_in_qdq_selection;
+};
```

Comment:
Can you clarify how a vector of DQ nodes across all the QDQ node groups is used?

I would have expected a way to find a NodeGroup from a node (at least from the target node, maybe from the others but not sure if that will be required) which target_node_to_qdq_group_ seems to provide, and a way to check if a node is in a  QDQNodeGroup for the check on whether it's supported or not in GetCapability). For the latter an unordered_set<const Node*> should be fine, with all the nodes including the target node.

Reply:
Yes. it makes more sense to retrieve the information for op support check from a NodeGroup obtained by a target node.

But since the `NodeGroup` is now represented as a vector of NodeIndexes, and we need to obtain the actual node, i.e. dq nodes in op_support_checker at least for conv operator to get the weight initializer. And it seems like given a NodeIndex, the way to get the node will be through the graph_viewer. 

Starting from `IsOpSupported()` method in getcapability(), it does not carry a graph_viewer param any more due to the expensive cost. So I have to retrieve nodes before the call to `isopsupported` and pass the nodes along. But now I realized this step is done much earlier as I changed to store the information in params struct. 

Sorry this dq_nodes vector was brought out by my old ideas.. I will change to having several lines to convert the NodeIndex->Node type to represent the qdq group and store that in the target_node_to_qdq_group_ and pass into params. I guess it sounds better?